src/Tuura/Plato/Translation.hs

```diff
@@ -76,8 +78,30 @@ validateInterface signs circuit
   where
     unused       = filter ((==Unused) . interface circuit) signs
 
-cartesianProduct :: NonEmpty.NonEmpty [a] -> [[a]]
-cartesianProduct l = sequence (NonEmpty.toList l)
+cartesianProduct :: Ord a => NonEmpty.NonEmpty [a] -> [[a]]
+cartesianProduct l = removeSupersets sortByLength
+  where
+    sequenced    = sequence (NonEmpty.toList l)
+    removeDupes  = map nub sequenced
+    removeNull   = filter (not . null) removeDupes
+    sortAllLists = map sort removeNull
+    sortByLength = sortBy (comparing length) sortAllLists
+
+removeSupersets :: Eq a => [[a]] -> [[a]]
+removeSupersets s = filter (not . null) result
+  where
+    prev n  = take n s
+    check x = checkForSupersets (s!!x) (prev x)
+    result  = map (\x -> if check x
+                        then []
+                        else s!!x
+                 )[0..(length s) - 1]
+
+checkForSupersets :: Eq a => [a] -> [[a]] -> Bool
+checkForSupersets current previous = any (`elem` subs) previous
```

Comment:
I think it will also be much faster, not slower!

* The current implementation needs to perform `elem` on each element of `x` in `previous`, which requires going through the list `subs` and comparing `x` with each element. If `subs` contains `m` elements, this will take `m * |x|` time in the worst case.

* My implementation needs to perform a single check `isSubsequenceOf` for each element of `x` in `previous`, which requires at most `|current| + |x|` time.

The latter is much faster, because `m` here contains all possible sublists of `current`, which means that `m = 2^|current|`!